.github/workflows/main.yaml

```diff
@@ -189,6 +226,66 @@ jobs:
           type: ondevice
           os: linux
 
+  # Runs on-host integration and unit tests.
+  on-device-test:
+    needs: [initialize, build]
+    if: |
+      needs.initialize.outputs.on_device_test == 'true' && (
+        inputs.nightly ||
+        inputs.postsubmit ||
+        contains(github.event.pull_request.labels.*.name, 'on_device')
+      )
+    runs-on: [self-hosted, linux, X64]
+    name: on_device_${{ matrix.name }}_${{ matrix.type }}
+    strategy:
+      fail-fast: false
+      matrix:
+        platform: ${{ fromJson(needs.initialize.outputs.platforms) }}
+        config: [devel]
+        type: ${{ fromJson(needs.initialize.outputs.on_device_test_tests) }}
+        include: ${{ fromJson(needs.initialize.outputs.includes) }}
+    container: ${{ needs.docker-unittest-image.outputs.docker_unittest_tag }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Install requirements
+        run: |
+          pip3 install grpcio==1.37.0 grpcio-tools==1.38.0
+        shell: bash
+      - name: Generate gRPC files
+        run: |
+          python -m grpc_tools.protoc -Itools/ --python_out=tools/ --grpc_python_out=tools/ tools/on_device_tests_gateway.proto
+        shell: bash
+      - name: Set up Cloud SDK
+        uses: google-github-actions/setup-gcloud@v0
+      - name: Set env vars
+        run: |
+          echo "PROJECT_NAME=$(gcloud config get-value project)" >> $GITHUB_ENV
+          echo "GITHUB_RUN_NUMBER=${GITHUB_RUN_NUMBER}" >> $GITHUB_ENV
+          echo "WORKFLOW=${{ github.workflow }}" >> $GITHUB_ENV
+          echo "LOGS_FILE_DIR=$GITHUB_WORKSPACE/out/tmp/${{ matrix.platform }}_${{ matrix.config }}_${{ matrix.type }}/" >> $GITHUB_ENV
+      - name: Create logs directory
+        run: |
+          mkdir -p "${LOGS_FILE_DIR}"
+        shell: bash
+      - name: run ${{ matrix.type }} tests on ${{ matrix.platform }} platform
+        id: on_device_test
+        run: |
```

Comment:
Can we limit GITHUB_TOKEN permissions for this job just to 'read', that would be enough to run this job right?

Reply:
So the only access we can restrict here is the `packages` access since all other scopes are restricted by the caller's workflow (Other wise the `permissions` will cause the job to fail). I sat up the job to have a `none` access on `packages`.

Reply:
Have you tried setting 'write' to none?  What happens if we do so?

Reply:
I don't think this is possible according to this document: https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs

What we can do is 
```YAML
permissions: read-all|write-all
```

If we use `permissions: read-all` the job will fail because it has `none` permission for everything except `packages`, so I toggled the `packages` permission from ~`write`~ to `none` since we don't really need it in this job.